### ISSUE TITLE
Regression: Infinite Loop between `Style/RedundantParentheses` and `Lint/AmbiguousRange`

### ISSUE THREAD

**Issue by jtannas** [rubocop/rubocop#12516]
Rubocop v1.58.0 introduces a regression that causes an infinite loop due to conflicting recommendations from  `Style/RedundantParentheses` and `Lint/AmbiguousRange`.

--------

## Expected behavior

I would expect one rule to take priority over the other so that no infinite loop occurs.

## Actual behavior
![image](https://github.com/rubocop/rubocop/assets/22785056/2a1e8204-b297-4958-ad5e-fd454019d2e7)
![image](https://github.com/rubocop/rubocop/assets/22785056/a64d1bbe-e31d-4a8a-9aa6-ca55427284be)

<details>
<summary>Stack Dump</summary>

```
186 files inspected, 4 offenses detected, 4 offenses corrected
Infinite loop detected in /home/jtannas/my-project/backend/app/models/foo.rb and caused by Lint/AmbiguousRange, Style/RedundantParentheses -> Lint/AmbiguousRange, Style/RedundantParentheses
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/lib/rubocop/runner.rb:332:in `check_for_infinite_loop'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/lib/rubocop/runner.rb:315:in `block in iterate_until_no_changes'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/lib/rubocop/runner.rb:314:in `loop'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/lib/rubocop/runner.rb:314:in `iterate_until_no_changes'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/lib/rubocop/runner.rb:283:in `do_inspection_loop'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/lib/rubocop/runner.rb:164:in `block in file_offenses'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/lib/rubocop/runner.rb:189:in `file_offense_cache'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/lib/rubocop/runner.rb:163:in `file_offenses'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/lib/rubocop/runner.rb:154:in `process_file'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/lib/rubocop/runner.rb:135:in `block in each_inspected_file'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/lib/rubocop/runner.rb:134:in `each'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/lib/rubocop/runner.rb:134:in `reduce'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/lib/rubocop/runner.rb:134:in `each_inspected_file'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/lib/rubocop/runner.rb:120:in `inspect_files'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/lib/rubocop/runner.rb:73:in `run'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/lib/rubocop/cli/command/execute_runner.rb:26:in `block in execute_runner'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/lib/rubocop/cli/command/execute_runner.rb:52:in `with_redirect'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/lib/rubocop/cli/command/execute_runner.rb:25:in `execute_runner'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/lib/rubocop/cli/command/execute_runner.rb:17:in `run'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/lib/rubocop/cli/command.rb:11:in `run'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/lib/rubocop/cli/environment.rb:18:in `run'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/lib/rubocop/cli.rb:118:in `run_command'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/lib/rubocop/cli.rb:125:in `execute_runners'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/lib/rubocop/cli.rb:51:in `block in run'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/lib/rubocop/cli.rb:77:in `profile_if_needed'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/lib/rubocop/cli.rb:43:in `run'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/exe/rubocop:19:in `block in <top (required)>'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/3.1.0/benchmark.rb:311:in `realtime'
/home/jtannas/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/rubocop-1.58.0/exe/rubocop:19:in `<top (required)>'
/home/jtannas/.rbenv/versions/3.1.2/bin/rubocop:25:in `load'
/home/jtannas/.rbenv/versions/3.1.2/bin/rubocop:25:in `<main>'
Finished in 1.370328764000078 seconds
```

</details>

## Steps to reproduce the problem

Sample code that triggers the issue
```rb
foo = [1, 2, 3, 4, 5]
end_indexes = [1, nil, 2, 3]
end_indexes.each do |end_index|
  puts foo[0..end_index || 0]
end
```

## RuboCop version

```
~/my-project (dependabot/bundler/rubocop-1.58.0 *)$ rubocop -V
1.58.0 (using Parser 3.2.2.4, rubocop-ast 1.30.0, running on ruby 3.1.2) [x86_64-linux]
  - rubocop-faker 1.1.0
  - rubocop-performance 1.19.1
  - rubocop-rails 2.22.2
  - rubocop-rspec 2.25.0
```

**Comment by koic** [comment]
This issue is a dup with https://github.com/rubocop/rubocop/issues/12436.
